File: python-sdk/saastify_edge/transformations/engine.py
"""
Transformation Engine

Executes transformation pipelines by applying a sequence of operations to values.
Supports both structured pipeline definitions and DSL string format.
"""

import re
from typing import Any, List, Dict, Optional, Tuple, Callable
from ..core.types import TransformationStep, RejectRow
from . import operations


# Build the transformation registry
TRANSFORMS: Dict[str, Callable] = {
    # Text operations
    "uppercase": operations.uppercase,
    "lowercase": operations.lowercase,
    "strip": operations.strip,
    "title_case": operations.title_case,
    "capitalize": operations.capitalize,
    "split_comma": operations.split_comma,
    "split": operations.split,
    "join": operations.join,
    "replace": operations.replace,
    "replace_regex": operations.replace_regex,
    "prefix": operations.prefix,
    "suffix": operations.suffix,
    "clean_html": operations.clean_html,
    "clean_upc": operations.clean_upc,
    
    # Numeric operations
    "clean_numeric_value": operations.clean_numeric_value,
    "addition": operations.addition,
    "subtraction": operations.subtraction,
    "multiplication": operations.multiplication,
    "division": operations.division,
    "percentage": operations.percentage,
    "adjust_negative_to_zero": operations.adjust_negative_to_zero,
    "zero_padding": operations.zero_padding,
    
    # Date operations
    "date_only": operations.date_only,
    
    # Control operations
    "set": operations.set_value,
    "set_number": operations.set_number,
    "copy": operations.copy,
    "rejects": operations.rejects,
    
    # Lookup operations
    "vlookup_map": operations.vlookup_map,
}


def apply_transformations(
    val: Any,
    steps: List[TransformationStep],
    context: Optional[Dict[str, Any]] = None
) -> Tuple[Any, bool]:
    """
    Apply a sequence of transformation steps to a value.
    
    Args:
        val: The input value
        steps: List of transformation steps with name and params
        context: Optional context (full row data, template info, etc.)
    
    Returns:
        Tuple of (transformed_value, is_rejected)
    """
    v = val
    
    for step in steps:
        func_name = step.get("name")
        if func_name not in TRANSFORMS:
            raise ValueError(f"Unknown transformation: {func_name}")
        
        func = TRANSFORMS[func_name]
        params = step.get("params", {})
        
        try:
            v = func(v, **params)
        except RejectRow:
            return None, True
        except Exception as e:
            # Log error but continue with original value
            print(f"Transformation error in {func_name}: {e}")
            continue
    
    return v, False


def transform(value: Any, rule_string: str) -> Any:
    """
    Apply DSL transformation rule to a single value (convenience function).
    
    Args:
        value: Input value to transform
        rule_string: DSL rule string (e.g., "strip + uppercase")
        
    Returns:
        Transformed value
        
    Example:
        >>> transform("  hello  ", "strip + uppercase")
        'HELLO'
    """
    results = bulk_apply_pipe_rules([value], rule_string)
    return results[0] if results else None


def bulk_apply_pipe_rules(
    values_list: List[Any],
    rule_strings: Any
) -> List[Any]:
    """
    Apply DSL transformation rules to a list of values.
    
    Supports broadcasting:
    - Single rule → multiple values
    - Multiple rules → single value  
    - Equal length arrays
    
    DSL Syntax:
    - Chain operations with ' + ' (spaces required)
    - Parameterize with '|': operation|param1|param2
    - Special delimiters:
      - split||| or split|\\| → split by pipe character
      - split|| → split by space
      - join|, → join with comma
    - vlookup|key1:val1,key2:val2
    - Arithmetic: addition|5, multiplication|2.5
    
    Args:
        values_list: List of values to transform
        rule_strings: String rule or list of rules
    
    Returns:
        List of transformed values
    """
    if not isinstance(values_list, list):
        raise ValueError("values_list must be a list")
    
    # Broadcasting logic
    if isinstance(rule_strings, str):
        rule_strings = [rule_strings] * len(values_list)
    elif isinstance(rule_strings, list):
        if len(rule_strings) == 1 and len(values_list) > 1:
            rule_strings *= len(values_list)
        elif len(values_list) == 1 and len(rule_strings) > 1:
            values_list *= len(rule_strings)
        elif len(rule_strings) != len(values_list):
            raise ValueError("Length mismatch between values_list and rule_strings")
    else:
        raise ValueError("rule_strings must be str or list[str]")
    
    results = []
    
    for val, rule in zip(values_list, rule_strings):
        if not rule:
            results.append(val)
            continue
        
        steps = _parse_dsl_rule(rule)
        out, _ = apply_transformations(val, steps)
        results.append(out)
    
    return results


def _parse_dsl_rule(rule: str) -> List[TransformationStep]:
    """
    Parse a DSL rule string into transformation steps.
    
    Args:
        rule: DSL rule string (e.g., "uppercase + strip + replace|old|new")
    
    Returns:
        List of TransformationStep dictionaries
    """
    steps: List[TransformationStep] = []
    
    # Split by ' + ' to get individual operations
    tokens = rule.replace(' + ', '|;|').split('|;|')
    
    for token_raw in tokens:
        token = token_raw.lstrip()
        if not token:
            continue
        
        # ─── Handle split operation ───
        if token.startswith("split|"):
            raw = token[6:]
            
            # Handle special delimiter cases
            if raw in ("||", r"\|"):
                delim = "|"
            elif raw.startswith("|||"):
                delim = raw[3:] or "|"
            elif raw.startswith("||"):
                tail = raw[2:]
                delim = " " if tail in ("", " ") else tail
            elif raw.startswith(r"\|"):
                delim = "|" + raw[2:]
            else:
                delim = raw
            
            if delim != " ":
                delim = delim.strip()
            if delim == "":
                raise ValueError("split requires non-empty delimiter")
            
            steps.append({"name": "split", "params": {"delimiter": delim}})
            continue
        
        # ─── Handle join operation ───
        if token.startswith("join|"):
            steps.append({"name": "join", "params": {"delimiter": token[5:]}})
            continue
        
        # ─── Handle prefix operation ───
        if token.startswith("prefix|"):
            steps.append({"name": "prefix", "params": {"prefix_str": token[7:]}})
            continue
        
        # ─── Handle suffix operation ───
        if token.startswith("suffix|"):
            steps.append({"name": "suffix", "params": {"suffix_str": token[7:]}})
            continue
        
        # ─── Handle replace_regex operation ───
        if token.startswith("replace_regex|"):
            parts = token[14:].split("||", 1)
            pat = parts[0]
            rep = parts[1] if len(parts) > 1 else ""
            steps.append({"name": "replace_regex", "params": {"pattern": pat, "repl": rep}})
            continue
        
        # ─── Handle replace operation ───
        if token.startswith("replace|"):
            try:
                old, new = token[8:].split('|', 1)
                steps.append({"name": "replace", "params": {"old": old, "new": new}})
            except ValueError:
                raise ValueError(f"replace requires two parameters: {token}")
            continue
        
        # ─── Handle vlookup operation ───
        if token.startswith("vlookup|"):
            mapping = {}
            for pair in token[8:].rstrip('|').split(','):
                k, v = pair.split(':', 1)
                v = v.strip()
                # Type coercion for values
                if v.lower() == "true":
                    v = True
                elif v.lower() == "false":
                    v = False
                elif re.match(r'^-?\d+\.\d+$', v):
                    v = float(v)
                elif v.isdigit():
                    v = int(v)
                
                mapping[k.strip().lower()] = v
            
            steps.append({"name": "vlookup_map", "params": {"mapping": mapping}})
            continue
        
        # ─── Handle arithmetic operations ───
        arithmetic_ops = {
            "addition": "amount",
            "subtraction": "amount",
            "multiplication": "factor",
            "division": "divisor",
            "percentage": "factor"
        }
        
        handled = False
        for op, param_name in arithmetic_ops.items():
            if token.startswith(f"{op}|"):
                param_value = token.split('|', 1)[1]
                steps.append({"name": op, "params": {param_name: param_value}})
                handled = True
                break
        
        if handled:
            continue
        
        # ─── Handle parameterized operations ───
        if token.startswith("strip|"):
            steps.append({"name": "strip", "params": {"chars": token.split('|', 1)[1]}})
        elif token.startswith("set|"):
            steps.append({"name": "set", "params": {"value": token.split('|', 1)[1]}})
        elif token.startswith("set_number|"):
            steps.append({"name": "set_number", "params": {"value": token.split('|', 1)[1]}})
        elif token.startswith("zero_padding|"):
            steps.append({"name": "zero_padding", "params": {"value": token.split('|', 1)[1]}})
        elif token == "rejects":
            steps.append({"name": "rejects"})
        else:
            # Simple operation without parameters
            steps.append({"name": token})
    
    return steps

File: js-sdk/src/transformations/engine.ts
/**
 * Transformation Engine - DSL parser and executor
 * Modular, easy to extend and debug
 */

import { TransformationStep, TransformFn, RejectRow } from '../core/types';
import * as text from './text';
import * as string from './string';
import * as numeric from './numeric';
import * as date from './date';
import * as list from './list';
import * as conditional from './conditional';
import * as utility from './utility';

// Registry of all transformation functions
const TRANSFORMS: Record<string, TransformFn> = {
  // Text operations
  uppercase: text.uppercase,
  lowercase: text.lowercase,
  strip: text.strip,
  title_case: text.titleCase,
  capitalize: text.capitalize,
  remove_whitespace: text.removeWhitespace,
  truncate: text.truncate,
  pad_left: text.padLeft,
  pad_right: text.padRight,
  reverse_string: text.reverseString,
  word_count: text.wordCount,
  char_count: text.charCount,
  extract_numbers: text.extractNumbers,
  extract_letters: text.extractLetters,
  remove_accents: text.removeAccents,
  remove_special_chars: text.removeSpecialChars,
  
  // String operations
  split: string.split,
  split_comma: string.splitComma,
  join: string.join,
  replace: string.replace,
  replace_regex: string.replaceRegex,
  prefix: string.prefix,
  suffix: string.suffix,
  slugify: string.slugify,
  to_snake_case: string.toSnakeCase,
  to_camel_case: string.toCamelCase,
  to_pascal_case: string.toPascalCase,
  sanitize_filename: string.sanitizeFilename,
  
  // Numeric operations
  clean_numeric_value: numeric.cleanNumericValue,
  addition: numeric.addition,
  subtraction: numeric.subtraction,
  multiplication: numeric.multiplication,
  division: numeric.division,
  percentage: numeric.percentage,
  round_decimal: numeric.roundDecimal,
  absolute_value: numeric.absoluteValue,
  ceiling: numeric.ceiling,
  floor: numeric.floor,
  square_root: numeric.squareRoot,
  power: numeric.power,
  modulo: numeric.modulo,
  clamp: numeric.clamp,
  scale: numeric.scale,
  reciprocal: numeric.reciprocal,
  sign: numeric.sign,
  adjust_negative_to_zero: numeric.adjustNegativeToZero,
  zero_padding: numeric.zeroPadding,
  
  // Date operations
  date_only: date.dateOnly,
  format_date: date.formatDate,
  add_days: date.addDays,
  subtract_days: date.subtractDays,
  day_of_week: date.dayOfWeek,
  day_name: date.dayName,
  month_name: date.monthName,
  year: date.year,
  month: date.month,
  day: date.day,
  is_weekend: date.isWeekend,
  days_between: date.daysBetween,
  
  // List operations
  list_length: list.listLength,
  list_first: list.listFirst,
  list_last: list.listLast,
  list_unique: list.listUnique,
  list_sort: list.listSort,
  
  // Conditional operations
  if_empty: conditional.ifEmpty,
  if_null: conditional.ifNull,
  coalesce: conditional.coalesce,
  copy: conditional.copy,
  rejects: conditional.rejects,
  set: conditional.set,
  set_number: conditional.setNumber,
  
  // Utility operations
  clean_html: utility.cleanHtml,
  clean_upc: utility.cleanUpc,
  vlookup_map: utility.vlookupMap,
  url_encode: utility.urlEncode,
  url_decode: utility.urlDecode,
  base64_encode: utility.base64Encode,
  base64_decode: utility.base64Decode,
  md5_hash: utility.md5Hash,
  json_parse: utility.jsonParse,
  json_stringify: utility.jsonStringify,
  xml_escape: utility.xmlEscape,
  html_unescape: utility.htmlUnescape,
  currency_format: utility.currencyFormat,
  levenshtein_distance: utility.levenshteinDistance,
  string_similarity: utility.stringSimilarity,
  extract_domain: utility.extractDomain,
};

/**
 * Parse a DSL rule string into transformation steps
 * Example: "strip + uppercase + replace| |_"
 */
export function parseRuleString(ruleString: string): TransformationStep[] {
  const steps: TransformationStep[] = [];
  
  // Split by " + " to get individual operations
  const operations = ruleString.split(' + ');
  
  for (const op of operations) {
    const parts = op.split('|');
    const operation = parts[0].trim();
    const args = parts.slice(1);
    
    steps.push({
      operation,
      args: args.reduce((acc, arg, i) => {
        acc[`arg${i}`] = arg;
        return acc;
      }, {} as Record<string, any>),
    });
  }
  
  return steps;
}

/**
 * Apply a sequence of transformation steps to a value
 */
export function applyTransformations(value: any, steps: TransformationStep[]): any {
  let result = value;
  
  for (const step of steps) {
    const fn = TRANSFORMS[step.operation];
    
    if (!fn) {
      console.warn(`Unknown transformation: ${step.operation}`);
      continue;
    }
    
    try {
      // Extract arguments in order
      const args = Object.keys(step.args)
        .sort()
        .map((key) => step.args[key]);
      
      result = fn(result, ...args);
    } catch (error) {
      if (error instanceof RejectRow) {
        throw error; // Re-throw RejectRow errors
      }
      console.error(`Error in transformation ${step.operation}:`, error);
      // Continue with current value on error
    }
  }
  
  return result;
}

/**
 * Transform a single value using a rule string
 * Convenience function for simple use cases
 */
export function transform(value: any, ruleString: string): any {
  const steps = parseRuleString(ruleString);
  return applyTransformations(value, steps);
}

/**
 * Bulk apply transformations with broadcasting
 * Supports: single rule → many values, many rules → single value, n:n
 */
export function bulkApplyPipeRules(
  valuesList: any[],
  ruleStrings: string | string[]
): any[] {
  const rules = Array.isArray(ruleStrings) ? ruleStrings : [ruleStrings];
  
  // Broadcasting logic
  if (rules.length === 1) {
    // Single rule applied to all values
    const steps = parseRuleString(rules[0]);
    return valuesList.map((v) => {
      try {
        return applyTransformations(v, steps);
      } catch (error) {
        if (error instanceof RejectRow) {
          return null; // Mark rejected rows as null
        }
        return v;
      }
    });
  }
  
  if (valuesList.length === 1) {
    // Multiple rules applied to single value
    return rules.map((rule) => {
      try {
        const steps = parseRuleString(rule);
        return applyTransformations(valuesList[0], steps);
      } catch (error) {
        if (error instanceof RejectRow) {
          return null;
        }
        return valuesList[0];
      }
    });
  }
  
  if (valuesList.length === rules.length) {
    // One-to-one mapping
    return valuesList.map((v, i) => {
      try {
        const steps = parseRuleString(rules[i]);
        return applyTransformations(v, steps);
      } catch (error) {
        if (error instanceof RejectRow) {
          return null;
        }
        return v;
      }
    });
  }
  
  throw new Error(
    `Invalid broadcasting: ${valuesList.length} values and ${rules.length} rules. ` +
    'Must be 1:n, n:1, or n:n.'
  );
}

/**
 * Get all available transformation operations
 */
export function getAvailableTransforms(): string[] {
  return Object.keys(TRANSFORMS).sort();
}

/**
 * Check if a transformation exists
 */
export function hasTransform(name: string): boolean {
  return name in TRANSFORMS;
}
